### PR TITLE
Copter:hexa - added suppposedly fault tolerant PPNNPN mixing

### LINKED ISSUE #7524: Copter: Additional Hex propeller configuration
Add hex propeller configuration that may be less susceptible to motor failure:

To quote @tuloski -
Basically changing the standard configuration of hexa with alternating propellers rotation PNPNPN, using a configuration PPNNPN can ensure that in 4 failures you keep the controllability, in 2 failures you don't. It is very interesting to implement probably. And btw in the two cases where you don't have full controllability, you can still loose yaw control, but land safely having the thrust, roll and pitch.

https://arxiv.org/pdf/1403.5986.pdf

This may require come creative coefficient selection to achieve the best results.

The propeller arrangement should be from front right CW, CW, CCW, CW, CCW, CCW

### PR DESCRIPTION
**Pull request by Vigasto**
Trying to fix #7524

Additional motor mixer configuration, but no actual tests yet. Configuration is used by using manual input on parameter lists.

Implementation might be changed if issues such as #908 is resolved

### CODE REVIEW

**Comment by peterbarker** [comment]
On Wed, 20 Mar 2019, Alif Ijlal W wrote:

> Trying to fix #7524
> 
> Additional motor mixer configuration, but no actual tests yet. Configuration is used by using manual input on parameter lists.
> 
> Implementation might be changed if issues such as #908 is resolved

I had a play around with this and autotest.

https://github.com/peterbarker/ardupilot/tree/hexa-ftc

The vehicle doesn't seem to get off the ground - BEFORE we fail a motor 
:-)

Would be really nice to have a working autotest to show this working 
before it is merged.

**Comment by Vigasto** [comment]
Thank you for the notice, I'll try to look into it

File: libraries/AP_Motors/AP_MotorsMatrix.cpp [hunk 1]
```diff
@@ -602,6 +602,24 @@ void AP_MotorsMatrix::setup_motors(motor_frame_class frame_class, motor_frame_ty
                     add_motor(AP_MOTORS_MOT_6,-150, AP_MOTORS_MATRIX_YAW_FACTOR_CW,  4);
                     success = true;
                     break;
+                case MOTOR_FRAME_TYPE_FTC_PLUS:
```
Comment (auturgy) [inline libraries/AP_Motors/AP_MotorsMatrix.cpp]: This is incorrect for + configuration.  You've described X configuration twice.

Comment (lthall) [reply libraries/AP_Motors/AP_MotorsMatrix.cpp]: I don't really think this is appropriate for a + configuration as it only really has symmetry as a x configuration. So I would simply not support the + config.

**Comment by lthall** [comment]
The mix is correct.

**Comment by muramura** [comment]
This pr's base source is old. I think you should update to the current source.

**Comment by OXINARF** [comment]
@Vigasto Do you still want to work on this?

**Comment by Vigasto** [comment]
Sorry for leaving this untouched for so long as I barely has time until recently. Is this okay?

**Comment by peterbarker** [comment]
@Vigasto You've pulled master into your branch.  You should have instead rebased on top of master.

Do you know how to fix that, or do you need some help?

**Comment by Vigasto** [comment]
...... yeah @peterbarker, that was some terrible history I pushed. I'd reset back before the merge happens and replay the changes manually since it was small, would that suffice? That will be two forced push in a row though

**Comment by peterbarker** [comment]
On Wed, 7 Aug 2019, Alif Ijlal W wrote:

> ...... yeah @peterbarker, that was some terrible history I pushed. I'd reset back before the merge happens and replay the changes manually since it was
> small, would that suffice? That will be two forced push in a row though

Whatever it takes; I tend to abuse our CI more than most :-)

Do my autotests pass on your branch ATM?

**Comment by Vigasto** [comment]
Yes, the autotests passed on my local machine just now

**Comment by IamPete1** [comment]
We now have a example this frame via scripting, https://github.com/ArduPilot/ardupilot/pull/16563

**Comment by rmtorre** [comment]
Hi guys,
Has anyone flew in this ppnnpn configuration? 
I have tried it yesterday. The drone was responding correctly in roll and pitch. But when I commanded yaw, it started to roll. Does anyone have some tips? 
Thanks in advance,
Rodrigo

**Pull request closed without merging**